lib/Dialects/VectorExt/Transform/VectorWarpUtils.cpp

```diff
@@ -173,6 +176,119 @@ struct WarpOpElementwise : public OpRewritePattern<WarpSingleLaneOp> {
+
+  // Reduce the two given scalars.
+  static Value reduceScalars(OpBuilder &b, Location loc, Value v1, Value v2,
+                             StringRef kind) {
+    if (kind == "add") {
+      if (v1.getType().isF32())
+        return b.create<arith::AddFOp>(loc, v1, v2);
+      return b.create<arith::AddIOp>(loc, v1, v2);
+    } else if (kind == "mul") {
+      if (v1.getType().isF32())
+        return b.create<arith::MulFOp>(loc, v1, v2);
+      return b.create<arith::MulIOp>(loc, v1, v2);
+    }
+    llvm_unreachable("unsupported reduction kind");
+  }
+
+  LogicalResult matchAndRewrite(WarpSingleLaneOp warpOp,
+                                PatternRewriter &rewriter) const override {
+    OpOperand *yieldOperand = getWarpResult(warpOp, [](Operation *op) {
+      return isa<vector::ReductionOp>(op);
+    });
+    if (!yieldOperand)
+      return failure();
+
+    auto reductionOp =
+        cast<vector::ReductionOp>(yieldOperand->get().getDefiningOp());
+    auto vectorType = reductionOp.vector().getType().cast<VectorType>();
+    // Only rank 1 vectors supported.
+    if (vectorType.getRank() != 1)
+      return failure();
+    // Only kWarpSize vectors supported.
+    if (vectorType.getShape()[0] != kWarpSize)
+      return failure();
+    // Only f32 and i32 element types are supported.
+    if (!reductionOp.getType().isF32() &&
+        !reductionOp.getType().isSignlessInteger(32))
+      return failure();
+    // Only add and mul supported.
+    if (reductionOp.kind() != "add" && reductionOp.kind() != "mul")
+      return failure();
+
+    // Define helper constants.
+    Location yieldLoc = yieldOperand->getOwner()->getLoc();
+    Value c0 = rewriter.create<arith::ConstantOp>(
+        yieldLoc, rewriter.getI32IntegerAttr(0));
+    Value cWarpSize = rewriter.create<arith::ConstantOp>(
+        yieldLoc, rewriter.getI32IntegerAttr(kWarpSize));
+
+    // Return vector that will be reduced from the WarpSingleLaneOp.
+    unsigned operandIndex = yieldOperand->getOperandNumber();
+    SmallVector<Value> yieldValues;
+    SmallVector<Type> retTypes;
+    yieldValues.push_back(reductionOp.vector());
+    retTypes.push_back(VectorType::get({1}, reductionOp->getResultTypes()[0]));
+    WarpSingleLaneOp newWarpOp = moveRegionToNewWarpOpAndAppendReturns(
+        rewriter, warpOp, yieldValues, retTypes);
+
+    // Every lane has one scalar value. These should be reduced.
+    Value laneValVec = newWarpOp.getResult(warpOp.getNumResults());
+    Value laneVal =
+        rewriter.create<vector::ExtractOp>(yieldLoc, laneValVec, 0);
+
+    // Parallel reduction: Every thread reduces two values. The result is
+    // stored at the lower thread. Requires log_2(kWarpSize) many parallel
+    // reductions.
+    for (int i = kWarpSize/2; i > 0; i /= 2) {
+      Value offset = rewriter.create<arith::ConstantOp>(
+          reductionOp.getLoc(), rewriter.getI32IntegerAttr(i));
+      Value shuffled = rewriter.create<gpu::ShuffleOp>(
```

Comment:
These should have better builders so we don't need to explicitly create useless constants at all call sites.
Could you update and bump the LLVM commit ? 
Ideally they'd also have a static form but that may be too much to ask..